sql: better byte buffer escaping in cli

RaduBerinde opened the pull request:
This change improves the escaping of byte[] in the sql shell:
- we now escape using `%+q` which guarantees ASCII output (issue #4315).
- we escape all strings that have non-printable ASCII characters, not just
  those that are not UTF8.
  @petermattis

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/4354)

<!-- Reviewable:end -->

petermattis commented on cli/sql_util.go:
```diff
@@ -174,9 +173,18 @@ func formatVal(val interface{}) string {
+		printable := true
+		for _, b := range t {
+			// Printable ASCII characters are between 0x20 and 0x7E.
+			if b < 0x20 || b > 0x7E {
```
https://golang.org/pkg/strconv/#IsPrint?

petermattis replied:
Hmm, I guess that isn't quite the same.

tamird replied:
But isn't that what you want?

Also, one of the few functional programming opportunities in Go!
```go
printable := bytes.IndexFunc(t, func(r rune) bool {
  return !unicode.IsPrint(r)
}) == -1
```

tamird replied:
or 
```go
printable := len(bytes.TrimLeftFunc(t, unicode.IsPrint)) == 0
```

tamird replied:
Oh, there is also [`unicode.IsControl`](https://godoc.org/unicode#IsControl) which might be useful.

petermattis commented:
LGTM

bdarnell commented:
We don't want to limit ourselves to just ascii; we should support all the "normal" unicode characters. I think that's `unicode.IsPrint` (and not `strconv.IsPrint`).

RaduBerinde commented:
I believe `unicode.IsPrint` is actually used by `%q` to decide which characters to escape. Apparently those characters were still a problem (see #4315)

---







*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4354#-:-KAMBcxlM3FW8VUanBOr:-641652161)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Can we figure out which character is a problem? We need to be able to add tests for this. @spencerkimball

RaduBerinde commented:
Wouldn't it depend on what terminal you are running the shell in?

If we want to support unescaped unicode characters, we also need to fix the alignment code (probably needs to use `RuneCountInString` instead of `len` somewhere)

---







*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4354#-:-KAMEMrYJ8FVatycHfRy:-1527422597)*
<!-- Sent from Reviewable.io -->

tamird commented:
Looks like it's any control character, and that IsPrint does the right thing.

I'll send a PR.

bdarnell commented:
Yes, it depends on the terminal (or, I believe, the `LANG` environment variable. I have `LANG=en_US.UTF-8` which enables unicode output in many applications). I think in 2016 it's better to default to utf-8 than ascii-only. 

`RuneCountInString` isn't enough because of combining characters. 

If `IsPrint` does the right thing and `%q` uses `IsPrint`, how did the problematic characters slip through in the first place?

tamird commented:
That `%q` does or doesn't do the right thing is irrelevant; if `utf8.Valid` returns true, we're just printing the byte array as a string.

RaduBerinde commented:
Tamir, if you look in #4315, the problematic is not `utf8.Valid` and it *is* quoted. The issue was actually not caused by my change to sometimes use `%s`.

I checked the source for fmt and `%q` uses https://golang.org/pkg/strconv/#Quote which uses https://golang.org/src/strconv/isprint.go

`unicode.IsPrint` is a different implementation, though it's not clear to me how the result differs exactly
https://golang.org/src/unicode/graphic.go?h=IsPrint#L50

---







*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4354#-:-KAMG8ERscLvvN0IDdxK:1578136480)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
*problematic string

---







*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4354#-:-KAMH9bTb8kz-RcNA0dn:-738173046)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
In #4315 the strings were not valid UTF8 and were getting printed with `%q`.

Looking at #4315 again I recognize the problem character: good old 'RIGHT-TO-LEFT OVERRIDE' (U+202E). It's apparently classified as a printable character because it is it is needed for correct display of mixed RTL and LTR strings. I'm not sure of the right way to exclude these. (Or we could let them through and append a 'POP DIRECTIONAL FORMATTING' (U+202C) to any string that may contain them)

RaduBerinde commented:
^that explains a lot of the weirdness I saw even in the browser when trying to copy/paste that string

---







*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4354#-:-KAMJ9tmu1IE0zEXJCDp:858408585)*
<!-- Sent from Reviewable.io -->

tamird commented:
Seems like the only thing we could check for (aside from this character verbatim) is the Cf category (https://godoc.org/unicode#pkg-variables) using unicode.In. I haven't been able to write a satisfying test for this, though.

RaduBerinde commented:
Updated. We now check `IsPrint` and use `%+q`. Used Tamir's suggestion (TrimLeftFunc). I'm still checking `utf8.Valid` because I'm not sure what `TrimLeftFunc`does if the string is not valid UTF-8, if someone knows for sure I can remove it.

I think this is decently safe, the only problem would be a string that is valid UTF-8 but contains some control characters that are not "matched" (like Ben mentioned above: RIGHT-TO-LEFT OVERRIDE without POP DIRECTIONAL FORMATTING).

---

Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4354#-:-KAMT1uZBbPDEfaR_c8M:1044736363)*
<!-- Sent from Reviewable.io -->

tamird commented:
TrimLeftFunc doesn't have special handling, but the rune it will pass to IsPrint will be `RuneError = '\uFFFD'` (https://godoc.org/unicode/utf8#pkg-constants), which presumably will cause IsPrint to return false.

Do you think we should also fall back to ascii-only if we find characters from the group Cf?

---

Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4354#-:-KAMUGKwt8b2m9UqnEkd:-823338018)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
I don't know enough about unicode to have a strong opinion, I searched a bit and there are a bunch of characters in Cf that seem harmless..

---







*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4354#-:-KAMY9M29Sy6YPajCyGP:612397873)*
<!-- Sent from Reviewable.io -->

bdarnell commented on cli/cli_test.go:
```diff
@@ -575,6 +575,33 @@ func Example_sql() {
+	//  - a valid printable UTF8 string using escapes
+	//  - a valid UTF8 string with a non-printable character
+	//  - a non-UTF8 string
+	c.RunWithArgs([]string{"sql", "-e", "insert into t.t values ('κόσμε'), (e'\\xc3\\xb1'), (e'\\x01'), (e'\\xc3\\x28')"})
```
Add `\u202e` (RTL override) as a test case (can we enter test cases with `\u` here or do we have to manually convert it to utf8? If so it's `\xe2\x80\xae`)

tamird replied:
the values are encoded according to http://www.postgresql.org/docs/current/static/datatype-binary.html#AEN5667 so I think it would be `e'\\xe280ae'`, though when I tried that test locally I didn't get the kind of junk I was expecting given the output in #4315 (even without `%+q`).

bdarnell commented:
Some of `Cf` may be harmless, but it's got a lot of invisible characters and when those are present it's probably best to fall back to the escaped format. (And contrary to what I said earlier, `unicode.IsPrint` excludes the Cf group, so I think this code should be fine; we just need to make sure the test covers the characters that motivated this issue).

RaduBerinde commented:
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[cli/cli_test.go, line 588 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4354#-KAMSs5JmqPr_2v86MkE-r2-588:-KAa0FCZ7BGXYfUo6cNq:531252437) ([raw file](https://github.com/cockroachdb/cockroach/blob/d864dcdc5b700a220e458f95c55772d9ff9e5ad8/cli/cli_test.go#L588)):*
I looked at the string in the issue more closely. This is the exact string:
https://play.golang.org/p/qq9OOsdJSw

First, note that using "+q" does help.

The problematic character is `DC 88` aka `U0708` (SYRIAC SUPRALINEAR COLON SKEWED LEFT which is right-to-left arabic): https://play.golang.org/p/1gcyI4MFVZ

Unfortunately this is considered printable by `IsPrint`.

So - the `%+q` change would have fixed the issue because we weren't unlucky enough for the binary blob to look like UTF-8. We still have a problem when the string does look like UTF-8. We have three solutions:
 - allow only printable ASCII (definite fix, but won't allow valid UTF8 strings)
 - allow printable UTF8, which may still exhibit the issue for some strings (unlikely for large enough arbitrary binary blobs, but still possible)
 - allow only a specific subset of UTF8, no idea how to go about deciding this subset

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4354)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[cli/cli_test.go, line 588 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4354#-KAMSs5JmqPr_2v86MkE-r2-588:-KAa7XAJNTgi_XygsbJe:1809143027) ([raw file](https://github.com/cockroachdb/cockroach/blob/d864dcdc5b700a220e458f95c55772d9ff9e5ad8/cli/cli_test.go#L588)):*
> allow only printable ASCII (definite fix, but won't allow valid UTF8 strings)

This is what I'd go with for now. As long as we have the test cases in place, we should be able to safely move to option 3 if there's sufficient need.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4354)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[cli/cli_test.go, line 588 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4354#-KAMSs5JmqPr_2v86MkE-r2-588:-KAa7bOOEza6_YK83ht0:458365189) ([raw file](https://github.com/cockroachdb/cockroach/blob/d864dcdc5b700a220e458f95c55772d9ff9e5ad8/cli/cli_test.go#L588)):*
OK, let's go with printable ASCII for now. 

There should be an API somewhere to query for the direction property of a character so we could exclude RTL characters in option 3, but it's not in the standard library's unicode module so let's not worry about that for now.

There's also option 4, to correctly print bidi strings and insert direction-control characters at field boundaries. However, this is not something we'd want to do unless we know the terminal can handle it reasonably so we'd need an ascii-only mode in any case. 

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4354)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
*[cli/cli_test.go, line 588 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4354#-KAMSs5JmqPr_2v86MkE-r2-588:-KAaBlGvqwF9aP6bsubB:948813269) ([raw file](https://github.com/cockroachdb/cockroach/blob/d864dcdc5b700a220e458f95c55772d9ff9e5ad8/cli/cli_test.go#L588)):*
Ok, I'll go with that and add a test with this character. Should we make sure to use `%+q` with empty strings as well? (so they show up as `""` instead of nothing?)

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4354)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[cli/cli_test.go, line 588 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4354#-KAMSs5JmqPr_2v86MkE-r2-588:-KAaCKtiR6PLwAV-JRm7:-525261790) ([raw file](https://github.com/cockroachdb/cockroach/blob/d864dcdc5b700a220e458f95c55772d9ff9e5ad8/cli/cli_test.go#L588)):*
Maybe we should use `%q` in the normal case, so we always get quotes? I don't remember what the reasoning for wanting the quotes removed was.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4354)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[cli/cli_test.go, line 588 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4354#-KAMSs5JmqPr_2v86MkE-r2-588:-KAaC92iSGYCskzebR1B:124235250) ([raw file](https://github.com/cockroachdb/cockroach/blob/d864dcdc5b700a220e458f95c55772d9ff9e5ad8/cli/cli_test.go#L588)):*
If we print empty strings as `""` then we should also print whitespace-only strings as `" "`. But I don't think it's important and I'd just go with what's easiest, which is to allow empty and whitespace-only strings to be invisible.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4354)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
*[cli/cli_test.go, line 588 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4354#-KAMSs5JmqPr_2v86MkE-r2-588:-KAaCoT_KMQ2F6JaiBz2:124810396) ([raw file](https://github.com/cockroachdb/cockroach/blob/d864dcdc5b700a220e458f95c55772d9ff9e5ad8/cli/cli_test.go#L588)):*
Tamir, see #4243 - we had quotes around column names, type names, table names, etc which was pretty annoying.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4354)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
*[cli/cli_test.go, line 588 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4354#-KAMSs5JmqPr_2v86MkE-r2-588:-KAaYkNuXiP7D2zu72qs:1466578188) ([raw file](https://github.com/cockroachdb/cockroach/blob/d864dcdc5b700a220e458f95c55772d9ff9e5ad8/cli/cli_test.go#L588)):*
PR updated.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4354)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
LGTM

tamird commented:
LGTM

---

Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4354#-:-KAaa8vtAcxNBAdgIxoT:676834233)*
<!-- Sent from Reviewable.io -->

RaduBerinde committed:
sql: better byte buffer escaping in cli

This change improves the escaping of byte[] in the sql shell:
 - we now escape using `%+q` which guarantees ASCII output (issue #4315).
 - we escape all strings that have non-printable ASCII characters. We don't
   allow UTF-8 strings because some characters can cause problems (see the
   discussion in #4354).
